Library/Homebrew/cmd/fetch.rb

```diff
@@ -176,81 +218,114 @@ def run
-        opoo "Patch reports different sha256: #{e.expected}"
-        Homebrew.failed = true
-      end
+            until remaining_downloads.empty?
+              begin
+                finished_states = [:fulfilled, :rejected]
 
-      def retry_fetch?(formula)
-        @fetch_tries ||= Hash.new { |h, k| h[k] = 1 }
-        if args.retry? && (@fetch_tries[formula] < FETCH_MAX_TRIES)
-          wait = 2 ** @fetch_tries[formula]
-          remaining = FETCH_MAX_TRIES - @fetch_tries[formula]
-          what = Utils.pluralize("tr", remaining, plural: "ies", singular: "y")
+                finished_downloads, remaining_downloads = remaining_downloads.partition do |_, future|
+                  finished_states.include?(future.state)
+                end
 
-          ohai "Retrying download in #{wait}s... (#{remaining} #{what} left)"
-          sleep wait
+                finished_downloads.each do |downloadable, future|
+                  previous_pending_line_count -= 1
+                  $stdout.print Tty.clear_to_end
+                  $stdout.flush
+                  output_message.call(downloadable, future)
+                end
 
-          formula.clear_cache
-          @fetch_tries[formula] += 1
-          true
-        else
-          Homebrew.failed = true
-          false
-        end
-      end
+                previous_pending_line_count = 0
+                remaining_downloads.each do |downloadable, future|
+                  # FIXME: Allow printing full terminal height.
+                  break if previous_pending_line_count >= [concurrency, (Tty.height - 1)].min
+
+                  $stdout.print Tty.clear_to_end
+                  $stdout.flush
+                  previous_pending_line_count += output_message.call(downloadable, future)
+                end
+
+                if previous_pending_line_count.positive?
+                  $stdout.print Tty.move_cursor_up_beginning(previous_pending_line_count)
+                  $stdout.flush
+                end
 
-      def fetch_fetchable(formula)
-        formula.clear_cache if args.force?
+                sleep 0.05
+              rescue Interrupt
+                remaining_downloads.each do |_, future|
+                  # FIXME: Implement cancellation of running downloads.
+                end
 
-        already_fetched = formula.cached_download.exist?
+                if previous_pending_line_count.positive?
+                  $stdout.print Tty.move_cursor_down(previous_pending_line_count - 1)
+                  $stdout.flush
+                end
 
-        begin
-          download = formula.fetch(verify_download_integrity: false)
-        rescue DownloadError
-          retry if retry_fetch?(formula)
-          raise
+                raise
+              end
+            end
+          ensure
+            $stdout.print Tty.show_cursor
+            $stdout.flush
+          end
         end
 
-        return unless download.file?
+        download_queue.shutdown
```

Comment:
Do we maybe want this in an `ensure` block?

Reply:
If we ensure the shutdown, I think we end up waiting for all downloads even if `Ctrl-C` is pressed. This correlates to the `FIXME` about cancellation of running downloads.

Reply:
Docs seem to suggest `shutdown` doesn't wait: https://github.com/ruby-concurrency/concurrent-ruby/blob/master/docs-source/thread_pools.md#thread-pool-status-and-shutdown. So seems like something that should always run.

But yeah maybe we also want `wait_for_termination` and/or `kill` depending on the circumstances.

Reply:
I added `kill` for now with a `FIXME` to implement graceful download cancellation.